include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -1257,9 +1283,10 @@ class ThreadPoolTempl : public onnxruntime::concurrency::ExtendedThreadPoolInter
     // Increase the worker count if needed.  Each worker will pick up
     // loops to execute from the current parallel section.
     std::function<void(unsigned)> worker_fn = [&ps](unsigned par_idx) {
+      ThreadPoolWaiter<kIsHybrid ? 4 : 0> waiter{};
```

Comment:
So, this change means on non-Intel CPUs you removed the SpinPause and replaced it with a busy loop. Why?  How is it better?

Reply:
Not quite, if `SpinPause` is a proper nop, the loop will be optimized out due to no side-effects.

Reply:
Why was the change made? I understand you wanted to optimize ONNX Runtime's performance for a few kinds of Intel CPUs, but why do you need to remove the spinlock for the other CPUs?

Reply:
How about changing 
```
if constexpr (kMaxBackoff == 1) {
        onnxruntime::concurrency::SpinPause();
}
```
to
```
if constexpr (kMaxBackoff <= 1) {
        onnxruntime::concurrency::SpinPause();
}
```

Reply:
Hmm, this doesn't remove the spin lock. This is just revising how we wait between testing the lock. I.e:
```
while(try_locked) { waiter.wait(); }
```

The `waiter.wait()` is just to help improve the perf/power consumption of the spin lock. The spin lock itself is the `while(try_locked) { /* Wait however you like */ }`

The exact values where picked based on benchmarks

Reply:
Here you set the template parameter `kMaxBackoff` to zero, which means `waiter.wait()` is empty, is it right?

Reply:
Yeah, it's means no waiting.

Reply:
But, before your change, there was a SpinPause. Why did you remove it for the cases that hybrid is false?

Reply:
Benchmarks... The configs chosen are from essentially a brute-force search for best wait times.

Reply:
Sorry there isn't some better answer, these weren't chosen by intuition or for any reason other than "it seemed to perform best regarding power/tput/lat on a variety of machines"

Reply:
Did you test it on AMD CPUs or ARM CPUs?

Reply:
This PR's title says "Spend less time busy waiting."  But  I see a very different fact here.  You are increasing the busy waiting time for non Intel CPUs. What's the rational behind of this?

Reply:
> Did you test it on AMD CPUs or ARM CPUs?

AMD I believe, ARM I'm not sure

Reply:
> This PR's title says "Spend less time busy waiting." But I see a very different fact here. You are increasing the busy waiting time for non Intel CPUs. What's the rational behind of this?

Hmm? There is no Intel only behavior in this PR. On systems that implemented 'SpinPause', it **varies** the number pauses. On systems that don't the new 'waiter' struct has no affect. But either way, the title is born from the key aspect of the patch which is decreasing the number of iterations in the main worker loop by 2 orders of magnitude (which , FWIW is similar to what upstream EigenV3 has done).

Reply:
This PR's title is "Spend less time busy waiting." I see you make Intel hybrid CPUs spend less time on busy waiting, and do the opposite for AMD CPUs. Is it true? Do we have data to support this decision?  Before your change, the code was like:
```c++
 while(condition is false){
  _mm_pause();
}
```
You think on AMD CPUs the "_mm_pause()" should be deleted. You modified the code to :
```
 while(condition is false);
```
You claimed based on benchmark data the change improved ONNX Runtime's performance on AMD CPUs. Am I understanding it right?

Reply:
> This PR's title is "Spend less time busy waiting." I see you make Intel hybrid CPUs spend less time on busy waiting, and do the opposite for AMD CPUs. Is it true? Do we have data to support this decision? Before your change, the code was like:

No, hybrid is not Intel vs AMD (Unless I really misunderstood the code in Onnx).  My understanding is Hybrid is any hybrid machine irrelevant of vendor (i.e Intel Alderlake, Mac M1, etc...). Non-hybrid is again any non-hybrid machine (i.e Intel Skylake, AMD Zen3, Zhaoxin Centaur).

My understanding is any machine that has both E (efficiency) and P (performance) cores should appear as hybrid, and any machine with only one or the other shouldn't be. Although again, this is entirely unrelated to vendor.

The tunings I did for hybrid where on: Alderlake, Raptor lake, Meteor Lake and Lunar Lake.
The tunings on non hybrid where: Skylake, SPR, Haswell

> 
> ```c++
>  while(condition is false){
>   _mm_pause();
> }
> ```
> 
> You think on AMD CPUs the "_mm_pause()" should be deleted. You modified the code to :
> 
> ```
>  while(condition is false);
> ```
no.
> 
> You claimed based on benchmark data the change improved ONNX Runtime's performance on AMD CPUs. Am I understanding it right?

I didn't tune for the AMD machines, although I just doubled checked, we did regression tests on an AMD machine and the patch had a positive impact.

Reply:
The is_hydrid_ flag is set [here](https://github.com/intel-sandbox/onnxruntime/blob/main/onnxruntime/core/common/cpuid_info.cc#L134) for x86 and here for [Arm](https://github.com/intel-sandbox/onnxruntime/blob/main/onnxruntime/core/common/cpuid_info.cc#L154). It doesn't look like is_hybrid_ flag is for Intel only.

Reply:
It's not about Intel vs AMD. I just used their CPUs as examples. I think we don't need to make a comprehensive list of hybrid/non-hybrid CPUs here. But we do not want to turn ONNX Runtime performance for a special kind of CPU, especially when the tunning may hurt performance for the other kinds of CPUs. I mentioned AMD because in my opinion this change is bad for all AMD CPUs. There is no good reason why we should change a spinlock to a busy loop for AMD CPUs. (You can argue it is not just hurting AMD CPUs, it hurts some other CPUs as well).   I am questioning why the " _mm_pause();"  function call was removed for most kind of CPUs, except Intel Hybrid CPUs.  I didn't mention ARM because as you said, you even didn't test it on ARM. Certainly there are a lot of other CPUs in the world as well.  I hope we do not need to discuss them case by case.

Reply:
> But we do not want to turn ONNX Runtime performance for a special kind of CPU, especially when the tunning may hurt performance for the other kinds of CPUs.

I would say the distinction between Hybrid and non-Hybrid is pretty critical in the context of a scheduler. It fundamentally changes the characteristics of the threads.

>  I mentioned AMD because in my opinion this change is bad for all AMD CPUs. There is no good reason why we should change a spinlock to a busy loop for AMD CPUs. (You can argue it is not just hurting AMD CPUs, it hurts some other CPUs as well). I am questioning why the " _mm_pause();" function call was removed for most kind of CPUs, except Intel Hybrid CPUs

Again, benchmarks indicates it was profitable (including on AMD). Look this single pause is not the end all be all. If as the maintainer you say "I want the pause", I will update the PR, but I don't think your reasoning thus far is sound.

> I didn't mention ARM because as you said, you even didn't test it on ARM

The changes the the wait code has no affect on arm. Its all a nop.

Reply:
My job is to find the deadlock. I commit I will finish it.  I am a maintainer of this repo, but I do not want to use it as authority to influence this technical discussion. I am not a specialist in thread scheduling. We will find someone who knows it better to review this code, after the deadlock bug is fixed.

Reply:
Thanks for your clarification. We'll discuss what you're saying.

Reply:
Regarding ARM, it is right that nop was used. There is a draft https://github.com/microsoft/onnxruntime/pull/22787 to add yield (no perf test is done yet).

Reply:
Just a guess, but a true yield would probably best be used far more sparingly than pause. Pause really exists to save a bit of power/prevent aggressive speculation, but it's orders of magnitude less costly than a yield. Especially if it's basically running at 1 thread per core, it seems if you actually want to go to the kernel you probably want to just be sleeping to save more power, using yield to bounce in and out of kernel space doesn't really accomplish that... But hey, Let's see what the benchmarks say